crates/driver/src/domain/eth/balances.rs

```diff
@@ -0,0 +1,95 @@
+use {
+    crate::{
+        domain::{competition::order, eth},
+        infra,
+    },
+    futures::future::join_all,
+    itertools::Itertools,
+    std::{collections::HashMap, sync::Arc},
+    tokio::sync::Mutex,
+};
+
+type BalanceGroup = (order::Trader, eth::TokenAddress, order::SellTokenBalance);
+type Balances = Arc<HashMap<BalanceGroup, order::SellAmount>>;
+
+#[derive(Default, Clone)]
+pub struct Cache(Arc<Mutex<Inner>>);
```

Comment:
nit: would it make sense to use RwLock here since once the first solver populates the balances there is no need to require the reading lock for the rest of the solvers?